### PR TITLE
NAS-111022 / 21.08 / Add validation related to legacy AFP shares

### PR DESCRIPTION
**Pull request by anodos325**
Since SMB shares will traverse mountpoints verify that new
share is not a child or parent of a share with different AFP
compatibility settings.

### CODE REVIEW

File: src/middlewared/middlewared/plugins/smb.py [hunk 1]
```diff
@@ -905,6 +907,7 @@ async def do_update(self, id, data):
         new['vuid'] = await self.generate_vuid(new['timemachine'], new['vuid'])
         await self.clean(new, 'sharingsmb_update', verrors, id=id)
         await self.validate(new, 'sharingsmb_update', verrors, old=old)
+        await self.legacy_afp_check(new, 'sharingsmb_verrors', verrors)
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/smb.py:910]: _update?

File: src/middlewared/middlewared/plugins/smb.py [hunk 2]
```diff
@@ -1073,6 +1076,28 @@ async def query(self, filters, options):
             return await super().query(filters, options)
         return result
 
+    @private
+    async def legacy_afp_check(self, data, schema, verrors):
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/smb.py:1080]: You could just return if data['afp'] is False?

Comment (anodos325) [reply src/middlewared/middlewared/plugins/smb.py:1080]: Not really, because an existing share on the same path may be `true`

Comment (william-gr) [reply src/middlewared/middlewared/plugins/smb.py:1080]: Sorry, if True

File: src/middlewared/middlewared/plugins/smb.py (hunk 1 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/smb.py:910]: The number of typos I make increases exponentially as the number of lines in diff decreases :(

**Review by rick-mesta — APPROVED** [review]
ship it 🚢

File: src/middlewared/middlewared/plugins/smb.py (hunk 2 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/smb.py:1080]: In that case, we may have a share on a parent or child directory that is False. :)

**Pull request merged**